node/test/Chain/Abstract.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE KindSignatures #-}
+
+-- | An abstract blockchain. This sits between the pure UTxO based DSL and a
+--   full-fat implementation of Cardano.
+
+module Chain.Abstract where
+
+import qualified Data.Foldable (fold, foldMap)
+import           Data.Hashable (hash)
+import qualified Data.Map.Strict as Map
+import           Data.Monoid (Sum (..))
+import qualified Data.Set as Set
+import           Pos.Core.Chrono
+import           Universum
+import qualified UTxO.DSL as DSL
+
+-- | In the abstract DSL, we identify transactions with integers.
+newtype Addr = Addr Int
+  deriving (Eq, Ord, Show)
+
+-- | Block hash
+newtype BlockHash = BlockHash Int
+
+-- | Hash of the genesis block
+genesisBlockHash :: BlockHash
+genesisBlockHash = BlockHash 0
+
+-- | Invalid block hash
+invalidBlockHash :: BlockHash
+invalidBlockHash = BlockHash $ Data.Hashable.hash @String "invalid"
+
+-- | Function with finite support. Note that the paper defines these as taking
+--   values in a Semiring, but there's nothing intrinsic to that constraint and
+--   Monoid is in the standard libraries.
+class Monoid v => FinitelySupportedFunction f k v | f -> k v where
+  fSupport :: f -> Set.Set k
+  -- | Apply the function. This must return `mempty` when the key is not in the support
+  -- of the function.
+  fApply :: f -> k -> v
+  fSum :: f -> v
+  fSum fn = Data.Foldable.foldMap (fApply fn) . fSupport $ fn
+
+-- | Standard implementation of a finitely supported function as a map.
+instance (Ord k, Monoid v) => FinitelySupportedFunction (Map.Map k v) k v where
+  fSupport = Map.keysSet
+  fApply m k = Map.findWithDefault mempty k m
+  fSum = Data.Foldable.fold . Map.elems
+
+data StakeDistribution a =
+  forall f. FinitelySupportedFunction f a (Sum Int) => StakeDistribution f
+
+data Repartition a =
+  forall f. FinitelySupportedFunction f a (Sum Int) => Repartition f
+
+data Delegation (h :: * -> *) a = Delegation
+  { delegator :: a
+  , delegatee :: a
+  }
+
+newtype SlotId = SlotId Int
+
+nextSlot :: SlotId -> SlotId
+nextSlot (SlotId i) = SlotId $ i + 1
+
+data Output (h :: * -> *) a = Output
+  { outAddr        :: a
+  , outVal         :: DSL.Value
+    -- | Repartitioning of the stake associated with this transaction's inputs.
+  , outRepartition :: Repartition a
+  }
+
+-- | Extract the DSL output from the abstract one.
+outDSL :: Output h a -> DSL.Output h a
+outDSL o = DSL.Output
+  { DSL.outAddr = outAddr o
+  , DSL.outVal = outVal o
+  }
+
+data Transaction h a = Transaction
+  { trFresh   :: DSL.Value
+  -- ^ The money that is created by this transaction. This money
+  -- implicitly comes from the treasury.
+  , trIns     :: NE (DSL.Input h a)
```

Comment:
If `Abstract.Transaction`'s `trIns` is a non-empty list, how do I convert a bootstrap DSL transaction into an abstract transaction? The bootstrap DSL transaction has an empty set of `DSL.Input`s.

Reply:
Such a translation has to fail, basically.